dagshub/models/model_locator.py

```diff
@@ -0,0 +1,320 @@
+import logging
+import os
+from dataclasses import dataclass
+from enum import auto, IntEnum
+from functools import cached_property
+from pathlib import Path, PurePosixPath
+from typing import Optional, Literal, Union, Tuple, List
+
+import dacite
+import yaml
+from pathvalidate import sanitize_filepath
+
+from dagshub.common.analytics import send_analytics_event
+from dagshub.common.api import RepoAPI
+from dagshub.common.api.repo import PathNotFoundError
+from dagshub.common.api.responses import StorageAPIEntry
+from dagshub.common.determine_repo import determine_repo
+from dagshub.common.helpers import log_message
+from dagshub.models.model_loaders import ModelLoader, RepoModelLoader, BucketModelLoader, DagsHubStorageModelLoader
+
+# TODO:
+"""
+Ways to get model:
+- automatic - with the cascade
+- from path
+- from bucket
+- from mlflow model
+- from mlflow artifact
+"""
+
+# TODO: don't forget to separate dvc downloads by the hash of the dir
+
+logger = logging.getLogger(__name__)
+
+
+class ModelNotFoundError(Exception):
+    def __str__(self):
+        return "Could not find model"
+
+
+@dataclass
+class ModelYamlStruct:
+    model_dir: str
+
+
+class StorageType(IntEnum):
+    Repo = auto()
+    DagshubStorage = auto()
+    Bucket = auto()
+
+
+class ModelLocator:
+    def __init__(
+        self,
+        repo: Optional[Union[str, RepoAPI]] = None,
+        path: Optional[str] = None,
+        bucket: Optional[str] = None,
+        git_ref: Optional[str] = None,
+        mlflow_model: Optional[str] = None,
+        mlflow_artifact: Optional[str] = None,
+        download_dest: Optional[Union[str, os.PathLike]] = None,
+        download_type: Literal["lazy", "eager"] = "eager",
+    ):
+        self._repo = repo
+        self.path = path
+        self.bucket = bucket
+        self._git_ref = git_ref
+        self.mlflow_model = mlflow_model
+        self.mlflow_artifact = mlflow_artifact
+        self._download_dest = download_dest
+        self.download_type = download_type
+
+    @cached_property
+    def repo_api(self) -> RepoAPI:
+        if isinstance(self._repo, RepoAPI):
+            return self._repo
+        if self._repo is None:
+            repo_api, git_ref = determine_repo()
```

Comment:
What if repo is passed as string what happens?

Reply:
It gets turned into `RepoAPI`, skipping both of the IFs: https://github.com/DagsHub/client/blob/aed3677a7c62bd36d2a599c8447ee166ef54aa9d/dagshub/models/model_locator.py#L82

The `RepoAPI` constructor specifically works with `user/repo` format